From ba0fef481ef1cb1169e3ed969ece32d62ae83f1e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:41 -0400
Subject: [powerpc] bpf: Fix the broken LD_VLAN_TAG_PRESENT test

Message-id: <1410545655-205645-593-git-send-email-dzickus@redhat.com>
Patchwork-id: 94656
O-Subject: [RHEL7 PATCH 592/626] powerpc: bpf: Fix the broken LD_VLAN_TAG_PRESENT test
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit dba63115ce0c888fcb4cdec3f8a4ba97d144afaf
Author: Denis Kirjanov <kda@linux-powerpc.org>
Date:   Wed Jun 25 21:34:57 2014 +0400

    powerpc: bpf: Fix the broken LD_VLAN_TAG_PRESENT test

    We have to return the boolean here if the tag presents
    or not, not just ANDing the TCI with the mask which results to:

    [  709.412097] test_bpf: #18 LD_VLAN_TAG_PRESENT
    [  709.412245] ret 4096 != 1
    [  709.412332] ret 4096 != 1
    [  709.412333] FAIL (2 times)

    Signed-off-by: Denis Kirjanov <kda@linux-powerpc.org>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/net/bpf_jit_comp.c b/arch/powerpc/net/bpf_jit_comp.c
index efea11d..c4a581f 100644
--- a/arch/powerpc/net/bpf_jit_comp.c
+++ b/arch/powerpc/net/bpf_jit_comp.c
@@ -400,10 +400,12 @@ static int bpf_jit_build_body(struct sk_filter *fp, u32 *image,
 
    PPC_LHZ_OFFS(r_A, r_skb, offsetof(struct sk_buff,
          vlan_tci));
-   if (filter[i].code == BPF_S_ANC_VLAN_TAG)
+   if (filter[i].code == BPF_S_ANC_VLAN_TAG) {
     PPC_ANDI(r_A, r_A, ~VLAN_TAG_PRESENT);
-   else
+   } else {
     PPC_ANDI(r_A, r_A, VLAN_TAG_PRESENT);
+    PPC_SRWI(r_A, r_A, 12);
+   }
    break;
   case BPF_S_ANC_QUEUE:
    BUILD_BUG_ON(FIELD_SIZEOF(struct sk_buff,
-- 
1.7.1